File: ruby-gem/lib/models/feature.rb
# frozen_string_literal: true

# Class from DFC Business Ontology: #Feature
require_relative '../core/semantic_object'


module DfcLinkmlConnector
  module Models
    class Feature < Core::SemanticObject
      SEMANTIC_TYPE = "dfc-b:Feature".freeze

      # @return [String]
      attr_accessor :date

      # @return [String]
      attr_accessor :description

      # @return [String]
      attr_accessor :name

      # @return [String]
      attr_accessor :characteristic_of

      # @return [String]
      attr_accessor :dimension

      # @return [Geometry]
      attr_accessor :geometry

      # @return [Array<Properties>]
      attr_accessor :properties

      # @param semanticId [String]
      # @param date: nil, description: nil, name: nil, characteristicOf: nil, dimension: nil, geometry: nil, properties: []
      def initialize(semanticId, date: nil, description: nil, name: nil, characteristicOf: nil, dimension: nil, geometry: nil, properties: [])
        super(semanticId)
        @date = date
        @description = description
        @name = name
        @characteristic_of = characteristicOf
        @dimension = dimension
        @geometry = geometry
        @properties = properties
        self.semanticType = "dfc-b:Feature"
        registerSemanticProperty("dfc-b:date", &method("date")).valueSetter = method("date=")
        registerSemanticProperty("dfc-b:description", &method("description")).valueSetter = method("description=")
        registerSemanticProperty("dfc-b:name", &method("name")).valueSetter = method("name=")
        registerSemanticProperty("dfc-b:characteristicOf", &method("characteristic_of")).valueSetter = method("characteristic_of=")
        registerSemanticProperty("dfc-b:hasDimension", &method("dimension")).valueSetter = method("dimension=")
        registerSemanticProperty("https://purl.org/geojson/vocab#geometry", &method("geometry")).valueSetter = method("geometry=")
        registerSemanticProperty("https://purl.org/geojson/vocab#properties", &method("properties")).valueSetter = method("properties=")
      end

      Core::SemanticObject.type_registry[SEMANTIC_TYPE] = self
    end
  end
end

File: typescript-connector/src/models/Feature.ts
// Class from DFC Business Ontology: #Feature
import { SemanticObject } from "../core/SemanticObject.js";
import type { Geometry } from "./Geometry.js";
import type { Properties } from "./Properties.js";

export interface FeatureParams {
  date?: string;
  description?: string;
  name?: string;
  characteristicOf?: string;
  hasDimension?: string;
  geometry?: Geometry;
  properties?: Properties[];
}
export class Feature extends SemanticObject {
  static get SEMANTIC_TYPE(): string {
    return "dfc-b:Feature";
  }

  date?: string;
  description?: string;
  name?: string;
  characteristicOf?: string;
  hasDimension?: string;
  geometry?: Geometry;
  properties?: Properties[];

  constructor(
    semanticId: string,
    params?: FeatureParams,
  ) {
    super(semanticId);
    this.date = params?.date;
    this.description = params?.description;
    this.name = params?.name;
    this.characteristicOf = params?.characteristicOf;
    this.hasDimension = params?.hasDimension;
    this.geometry = params?.geometry;
    this.properties = params?.properties;
    this.semanticType = Feature.SEMANTIC_TYPE;
    this.registerSemanticProperty("dfc-b:date", () => this.date);
    this.registerSemanticProperty("dfc-b:description", () => this.description);
    this.registerSemanticProperty("dfc-b:name", () => this.name);
    this.registerSemanticProperty("dfc-b:characteristicOf", () => this.characteristicOf);
    this.registerSemanticProperty("dfc-b:hasDimension", () => this.hasDimension);
    this.registerSemanticProperty("https://purl.org/geojson/vocab#geometry", () => this.geometry);
    this.registerSemanticProperty("https://purl.org/geojson/vocab#properties", () => this.properties);
  }
  static {
    SemanticObject.typeRegistry.set(Feature.SEMANTIC_TYPE, Feature);
  }
}
